File: frontend/scripts/build-public-media-projection.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const repoRoot = path.resolve(__dirname, "../..");
const distStaticDir = path.resolve(__dirname, "../dist/static/images");
const generatedDir = path.resolve(__dirname, "../generated");

fs.mkdirSync(generatedDir, { recursive: true });

console.log("[Projection Compiler] Loading canonical sources...");
const manifestPath = path.join(repoRoot, "data/images/sources/manifest.json");
const catManifestPath = path.join(repoRoot, "data/images/sources/category_manifest.json");
const rejectedPath = path.join(repoRoot, "data/images/sources/rejected_candidates.json");
const pubPath = path.join(repoRoot, "data/images/sources/publishability_report.json");

const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf-8"));
const categories = JSON.parse(fs.readFileSync(catManifestPath, "utf-8"));

const rejectedIds = new Set();
if (fs.existsSync(rejectedPath)) {
  const rejected = JSON.parse(fs.readFileSync(rejectedPath, "utf-8"));
  for (const r of rejected) {
    const rid = r.research_id || r.place_id;
    if (rid) rejectedIds.add(rid);
  }
}

const pubDecisions = {};
if (fs.existsSync(pubPath)) {
  const pub = JSON.parse(fs.readFileSync(pubPath, "utf-8"));
  for (const d of pub.decisions || []) {
    pubDecisions[d.place_id] = d;
  }
}

const PUBLIC_VARIANTS = ["hero.webp", "card.webp", "thumbnail.webp"];

const publicManifest = {
  generated_at: process.env.VITE_BUILD_TIME || new Date().toISOString(),
  compiler: "build-public-media-projection.js",
  rules: {
    allowed_variants: PUBLIC_VARIANTS,
    excluded_variants: ["original.webp"],
    verification_requirement: "EXACT_LOCATION_VERIFIED or VERIFIED_AUTHENTIC_PHOTOGRAPHY",
    rejected_policy: "NEVER_PUBLISH",
  },
  places: {},
  categories: {},
};

const filesToCopy = [];
let totalOriginalBytesSaved = 0;

for (const item of manifest) {
  const placeId = item.place_id;
  const assetHash = item.asset_hash;
  const vStatus = item.verification_status || "";

  if (rejectedIds.has(placeId)) continue;

  const d = pubDecisions[placeId] || {};
  if (d.classification === "RELATED_LOCATION_ONLY") continue;

  if (
    vStatus !== "VERIFIED_AUTHENTIC_PHOTOGRAPHY" &&
    vStatus !== "EXACT_LOCATION_VERIFIED" &&
    d.classification !== "EXACT_LOCATION_VERIFIED"
  ) {
    continue;
  }

  const srcPlaceDir = path.join(repoRoot, "data/images/places", placeId, assetHash);
  if (!fs.existsSync(srcPlaceDir)) continue;

  const destPlaceDir = path.join(distStaticDir, "places", placeId, assetHash);

  const placeEntry = {
    place_name: item.place_name,
    asset_hash: assetHash,
    title: item.title,
    alt_text: item.alt_text,
    creator: item.creator,
    license: item.license,
    attribution: item.attribution,
    variants: {},
  };

  const origFile = path.join(srcPlaceDir, "original.webp");
  if (fs.existsSync(origFile)) {
    totalOriginalBytesSaved += fs.statSync(origFile).size;
  }

  for (const v of PUBLIC_VARIANTS) {
    const srcF = path.join(srcPlaceDir, v);
    if (fs.existsSync(srcF)) {
      const destF = path.join(destPlaceDir, v);
      filesToCopy.push({ src: srcF, dest: destF });
      const vName = v.replace(".webp", "");
      placeEntry.variants[vName] = {
        path: `/static/images/places/${placeId}/${assetHash}/${v}`,
        size_bytes: fs.statSync(srcF).size,
      };
    }
  }

  publicManifest.places[placeId] = placeEntry;
}

for (const cat of categories) {
  const catId = cat.category_id;
  const assetHash = cat.asset_hash;
  const srcCatDir = path.join(repoRoot, "data/images/categories", catId, assetHash);
  const destCatDir = path.join(distStaticDir, "categories", catId, assetHash);

  const origFile = path.join(srcCatDir, "original.webp");
  if (fs.existsSync(origFile)) {
    totalOriginalBytesSaved += fs.statSync(origFile).size;
  }

  const catEntry = {
    category_key: cat.category_key,
    asset_hash: assetHash,
    title: cat.title,
    variants: {},
  };

  for (const v of PUBLIC_VARIANTS) {
    const srcF = path.join(srcCatDir, v);
    if (fs.existsSync(srcF)) {
      const destF = path.join(destCatDir, v);
      filesToCopy.push({ src: srcF, dest: destF });
      const vName = v.replace(".webp", "");
      catEntry.variants[vName] = {
        path: `/static/images/categories/${catId}/${assetHash}/${v}`,
        size_bytes: fs.statSync(srcF).size,
      };
    }
  }

  publicManifest.categories[catId] = catEntry;
}

const manifestOut = path.join(generatedDir, "publicMediaManifest.json");
fs.writeFileSync(manifestOut, JSON.stringify(publicManifest, null, 2), "utf-8");
console.log(
  `[Projection Compiler] Wrote public manifest with ${Object.keys(publicManifest.places).length} places, ${Object.keys(publicManifest.categories).length} categories -> ${manifestOut}`
);

const distDir = path.resolve(__dirname, "../dist");
if (fs.existsSync(distDir)) {
  if (fs.existsSync(distStaticDir)) {
    fs.rmSync(distStaticDir, { recursive: true, force: true });
  }

  let copiedCount = 0;
  let copiedBytes = 0;
  for (const { src, dest } of filesToCopy) {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.copyFileSync(src, dest);
    copiedCount++;
    copiedBytes += fs.statSync(dest).size;
  }

  console.log(
    `[Projection Compiler] Deployed ${copiedCount} public variant files (${(copiedBytes / (1024 * 1024)).toFixed(2)} MB) to ${distStaticDir}`
  );
  console.log(
    `[Projection Compiler] Successfully eliminated ${(totalOriginalBytesSaved / (1024 * 1024)).toFixed(2)} MB of unneeded original.webp sources!`
  );
}

File: scripts/build_public_media_projection.py
import json
import os
import shutil
import sys
import time

def main():
    repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
    dist_static_dir = os.path.join(repo_root, "frontend/dist/static/images")
    generated_dir = os.path.join(repo_root, "frontend/generated")
    os.makedirs(generated_dir, exist_ok=True)

    print("[Projection Compiler] Loading canonical sources...")
    manifest_path = os.path.join(repo_root, "data/images/sources/manifest.json")
    category_manifest_path = os.path.join(repo_root, "data/images/sources/category_manifest.json")
    rejected_path = os.path.join(repo_root, "data/images/sources/rejected_candidates.json")
    publishability_path = os.path.join(repo_root, "data/images/sources/publishability_report.json")

    with open(manifest_path, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    with open(category_manifest_path, "r", encoding="utf-8") as f:
        categories = json.load(f)

    rejected_ids = set()
    if os.path.exists(rejected_path):
        with open(rejected_path, "r", encoding="utf-8") as f:
            for r in json.load(f):
                rid = r.get("research_id") or r.get("place_id")
                if rid:
                    rejected_ids.add(rid)

    pub_decisions = {}
    if os.path.exists(publishability_path):
        with open(publishability_path, "r", encoding="utf-8") as f:
            for d in json.load(f).get("decisions", []):
                pub_decisions[d.get("place_id")] = d

    # Preferred public variants for production serving:
    # hero, card, thumbnail. (NEVER original.webp!)
    PUBLIC_VARIANTS = ["hero.webp", "card.webp", "thumbnail.webp"]

    public_manifest = {
        "generated_at": os.environ.get("VITE_BUILD_TIME") or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "compiler": "build_public_media_projection.py",
        "rules": {
            "allowed_variants": PUBLIC_VARIANTS,
            "excluded_variants": ["original.webp"],
            "verification_requirement": "EXACT_LOCATION_VERIFIED or VERIFIED_AUTHENTIC_PHOTOGRAPHY",
            "rejected_policy": "NEVER_PUBLISH"
        },
        "places": {},
        "categories": {}
    }

    files_to_copy = []
    total_original_bytes_saved = 0

    # 1. Process place media
    for item in manifest:
        place_id = item["place_id"]
        asset_hash = item.get("asset_hash")
        v_status = item.get("verification_status", "")

        # Skip rejected
        if place_id in rejected_ids:
            continue

        # Check classification
        d = pub_decisions.get(place_id, {})
        classification = d.get("classification")
        if classification == "RELATED_LOCATION_ONLY":
            # RELATED_LOCATION cannot be public HERO/CARD
            continue

        if v_status not in ["VERIFIED_AUTHENTIC_PHOTOGRAPHY", "EXACT_LOCATION_VERIFIED"] and classification != "EXACT_LOCATION_VERIFIED":
            continue

        src_place_dir = os.path.join(repo_root, f"data/images/places/{place_id}/{asset_hash}")
        if not os.path.exists(src_place_dir):
            print(f"[Projection Compiler] Warning: source directory missing for {place_id}/{asset_hash}")
            continue

        dest_place_dir = os.path.join(dist_static_dir, f"places/{place_id}/{asset_hash}")

        place_entry = {
            "place_name": item.get("place_name"),
            "asset_hash": asset_hash,
            "title": item.get("title"),
            "alt_text": item.get("alt_text"),
            "creator": item.get("creator"),
            "license": item.get("license"),
            "attribution": item.get("attribution"),
            "variants": {}
        }

        # Check for original.webp to track bytes saved
        orig_file = os.path.join(src_place_dir, "original.webp")
        if os.path.exists(orig_file):
            total_original_bytes_saved += os.path.getsize(orig_file)

        for v in PUBLIC_VARIANTS:
            src_f = os.path.join(src_place_dir, v)
            if os.path.exists(src_f):
                dest_f = os.path.join(dest_place_dir, v)
                files_to_copy.append((src_f, dest_f))
                v_name = v.replace(".webp", "")
                place_entry["variants"][v_name] = {
                    "path": f"/static/images/places/{place_id}/{asset_hash}/{v}",
                    "size_bytes": os.path.getsize(src_f)
                }

        public_manifest["places"][place_id] = place_entry

    # 2. Process category media
    for cat in categories:
        cat_id = cat["category_id"]
        asset_hash = cat["asset_hash"]
        src_cat_dir = os.path.join(repo_root, f"data/images/categories/{cat_id}/{asset_hash}")
        dest_cat_dir = os.path.join(dist_static_dir, f"categories/{cat_id}/{asset_hash}")

        # Check for original.webp to track bytes saved
        orig_file = os.path.join(src_cat_dir, "original.webp")
        if os.path.exists(orig_file):
            total_original_bytes_saved += os.path.getsize(orig_file)

        cat_entry = {
            "category_key": cat.get("category_key"),
            "asset_hash": asset_hash,
            "title": cat.get("title"),
            "variants": {}
        }

        for v in PUBLIC_VARIANTS:
            src_f = os.path.join(src_cat_dir, v)
            if os.path.exists(src_f):
                dest_f = os.path.join(dest_cat_dir, v)
                files_to_copy.append((src_f, dest_f))
                v_name = v.replace(".webp", "")
                cat_entry["variants"][v_name] = {
                    "path": f"/static/images/categories/{cat_id}/{asset_hash}/{v}",
                    "size_bytes": os.path.getsize(src_f)
                }

        public_manifest["categories"][cat_id] = cat_entry

    # Save publicMediaManifest.json
    manifest_out = os.path.join(generated_dir, "publicMediaManifest.json")
    with open(manifest_out, "w", encoding="utf-8") as f:
        json.dump(public_manifest, f, indent=2)
    print(f"[Projection Compiler] Wrote public manifest with {len(public_manifest['places'])} places, {len(public_manifest['categories'])} categories -> {manifest_out}")

    # Copy files into dist if dist exists
    dist_dir = os.path.join(repo_root, "frontend/dist")
    if os.path.exists(dist_dir):
        # Clean existing static/images directory in dist to guarantee no stale original.webp remain
        if os.path.exists(dist_static_dir):
            shutil.rmtree(dist_static_dir)

        copied_count = 0
        copied_bytes = 0
        for src_f, dest_f in files_to_copy:
            os.makedirs(os.path.dirname(dest_f), exist_ok=True)
            shutil.copy2(src_f, dest_f)
            copied_count += 1
            copied_bytes += os.path.getsize(dest_f)

        print(f"[Projection Compiler] Deployed {copied_count} public variant files ({round(copied_bytes / (1024*1024), 2)} MB) to {dist_static_dir}")
        print(f"[Projection Compiler] Successfully eliminated {round(total_original_bytes_saved / (1024*1024), 2)} MB of unneeded original.webp sources!")
    else:
        print("[Projection Compiler] Dist directory does not exist yet. Manifest generated; run after vite build.")

if __name__ == "__main__":
    main()
